<a href="https://colab.research.google.com/github/habsgirl19/Hockey_East_Data_Tracking/blob/main/Kristi_Heiman_Lesson_7_2_Python_Wrap_Up.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### 0. Python and Jupyter

0.2 Delete the extra cells under this cell.

0.1 Make a new cell under this one and add your name in it. Convert this cell to a markdown cell, and run the cell.

Kristi Heiman

0.3 Comment out (don't delete code) the correct lines so that this cell only prints "Only print this line!".

In [ ]:
#print("Comment out this line!")
#print("Don't print this.")
#print("Don't print this ether.")
print('Only print this line!')

Only print this line!


0.4 Add line numbers to the cell with code in 0.3.

### 1. Math

1.1 Use python to add 123456 and 654321.

In [ ]:
123456 + 654321

777777

1.2 Use python to calculate 2 to the 10th power.

In [ ]:
2**10

1024

### 2. Variables and types

2.1 Create a variable `r` and assign the number 10 to it.

In [ ]:
r = 10

2.2 Use the formula for the area of a circle and your variable `r` to calculate the area of a 10 inch pizza. If you don't know the formula for the area of a circle, **<font color = 'blue'> Google </font>** it. You can use 3.14 for the value of pi.

In [ ]:
#search formula to calculate the circumferece of a circle
#pi = 3.14
#r = float(input("Enter the radius of a circle:"))
#area - PI*r*r
#print("Area of a circe = %.2f"%area)

PI = 3.14
r = float(input("Enter the radius of a circle:"))
area = PI*r*r
print("Area of a circe = %.2f"%area)

r = 10
area = PI * 10 * 10


Enter the radius of a circle:10
Area of a circe = 314.00


2.3 copy your code from 2.2 and change the value of `r` to find the area of a 14 inch pizza.

In [ ]:
PI = 3.14
r = float(input("Enter the radius of a circle:"))
area = PI*r*r
print("Area of a circe = %.2f"%area)

r = 14
area = PI*r*r


Enter the radius of a circle:14
Area of a circe = 615.44


2.4 What is the type of the variable `x`?

In [ ]:
x = '3'
print(type(x))

<class 'str'>


2.5 Convert `x` into an int. 

In [ ]:
num = '3'

#check and print the type num variable
print(type(num))

#convert the num into a string

converted_num = int(num)
print(type(converted_num))


<class 'str'>
<class 'int'>


2.6 Print out the type of `x` to confirm that `x` in now an int.

In [ ]:
#print type of converted _num
#convert the num into a string

converted_num = int(num)
print(type(converted_num))

print(type(converted_num))

#check by doing a mathematical operation
print(converted_num + 20.5)

<class 'int'>
<class 'int'>
23.5


Make a variable `y`, assign it a value of 6 and multiply `x` and `y`

In [ ]:
y = 6 
y * x

'333333'

### 3. Strings

3.1 Create a variable called `name` and another variable called `favorite_pet` and fill them with the correct values for you. Print out both variables.

In [ ]:
name = 'Kristi'
favorite_pet = 'dog'
print(name, favorite_pet)

Kristi dog


3.2 Use string formatting to print the phrase "My name is _, and my favorite pet is _.", with the blanks filled in with the variables you created in 3.1.

In [ ]:
name = "Kristi"
favorite_pet = "dog"

print(f"My name is {name} and my favorite pet is a {favorite_pet}.")

My name is Kristi and my favorite pet is a dog.


3.3 Use sting indexing to print out just the word "fox" from `phrase`. Dont include any spaces.

In [ ]:
phrase = "The quick brown fox jumps over the lazy dog"
print(phrase.find("fox"))


16


3.4 How long is `phrase`?

In [ ]:
len(phrase)

43

### 4. Lists

4.1 Create a list called `evens` that has the numbers 2, 4, 6, and 8.

In [ ]:
evens = [2, 4, 6, 8]
print(evens)

[2, 4, 6, 8]


4.2 Add the number 10 to your list using the append function.

In [ ]:
evens.append(10)
print(evens)

[2, 4, 6, 8, 10]


4.3 Use indexing to print the second to last number in your list (it should be 8).

In [ ]:
evens[-2]

8

4.4 How long is your list?

In [ ]:
len(evens)

5

Delete the number 6 from your list. Print the length to confirm the list is now one shorter than it was previously

In [ ]:
evens.remove(6)
print(evens)

[2, 4, 8, 10]


### 5. Numpy

5.1 Import the library `numpy` with the alias `np`. 

In [ ]:
import numpy as np

5.2 Use numpy to find the minimum, maximum, and mean of `evens`.

In [ ]:
print(min(evens))


2


In [ ]:
print(max(evens))

10


In [ ]:
print(np.mean(evens))

6.0


5.3 Use numpy to print a more accurate estimate of pi. (**<font color = 'blue'> Google </font>** "numpy pi" if you get stuck, and check out the 2nd or 3rd link!)

In [ ]:
import numpy as np

pi_value = np.pi
print(pi_value)


3.141592653589793


5.4 Use numpy to find the sqare root of 2.

In [ ]:
arr = np.sqrt(2)
print(arr)

1.4142135623730951


### 6. Loops

6.1 use a for loop and `evens` to print:  
2 is even  
4 is even  
6 is even  
8 is even  
10 is even  

In [ ]:
evens = [2, 4, 6, 8, 10]

for num in evens:
  if num % 2 ==0:
   print(num, "is even")


  


2 is even
4 is even
6 is even
8 is even
10 is even


6.2 Modify this loop so that it prints out the first 20 perfect squares (1, 4, 9, ... 400).

In [ ]:
for iNumber in range(1,21):
    print(iNumber, '\t', iNumber **2)

1 	 1
2 	 4
3 	 9
4 	 16
5 	 25
6 	 36
7 	 49
8 	 64
9 	 81
10 	 100
11 	 121
12 	 144
13 	 169
14 	 196
15 	 225
16 	 256
17 	 289
18 	 324
19 	 361
20 	 400


### 7. if/else and booleans

7.1 In one or two sentences explain what you think this code is doing.

In [ ]:
letter = "w"
letter in "word"


True

The code in Question 7.1 implements a boolean (or logical expression) to evaluate whether or not the variable "letter" which stores the letter "w". After running the code, in this case a function, the code is excuted to return an answer of a boolean function. Here, it is true that "w" is in word.

7.2 What would the code from 7.1 print if you change line one to  
`letter = "u"`?

In [ ]:
letter = "u"
letter in "word"

False

7.3 make a variable `number` and set it equal to 6. Write an if-else statement that compares `number` to ten and ether prints "The number is smaller than 10." or "The number is grater than or equal to 10." Change the value of `number` to 12 and rerun your code to make sure the statment is working. (Bonus: make the statment print "The number is 10." when the number is equal to 10)

In [ ]:
number = 10

if number < 10:
  print("The number is smaller than 10.")
elif number == 10:
  print("The number is 10.")
else:
  print("The number is greater or equal to 10.")


The number is 10.


In [ ]:
number = 15

if number < 10:
  print("The number is smaller than 10.")
elif number == 10:
  print("The number is 10.")
else:
  print("The number is greater or equal to 10.")


The number is greater or equal to 10.


In [ ]:
number = -25

if number < 10:
  print("The number is smaller than 10.")
elif number == 10:
  print("The number is 10.")
else:
  print("The number is greater or equal to 10.")


The number is smaller than 10.


7.4 Build an if statment that checks if a number is even (hint: **<font color = 'blue'> Google </font>** is your friend, and you may need the % sign), check that your code works on 1, 2, 3, and 10

In [ ]:
x = 3466

if (x % 2)==0:
  print(x, "is even.")
if (x % 2)!=0:
    print(x, "is odd.")

3466 is even.


In [ ]:
x = 1
if (x % 2)!=0:
  print(x, 'is not even')
if (x % 2)==0:
  print(x, 'is even.')

1 is not even


7.5 Use a `for` loop and an `if` statment to print:  

1 is odd.  
2 is even.  
3 is odd.  
4 is even.  
5 is odd.  
6 is even.  
7 is odd.  
8 is even.  
9 is odd.  
10 is even.  
11 is odd.  
12 is even.  
13 is odd.  
14 is even.  
15 is odd.  

In [ ]:
for x in range(16):
  if (x % 2)==0:
    print(x, "is even.")
  if (x % 2)!=0:
    print(x, "is odd.")

0 is even.
1 is odd.
2 is even.
3 is odd.
4 is even.
5 is odd.
6 is even.
7 is odd.
8 is even.
9 is odd.
10 is even.
11 is odd.
12 is even.
13 is odd.
14 is even.
15 is odd.


### 8. Custom Functions

8.1 Write a python function called `encouragement` that returns a custom message to a list of students. The message should be "Keep going, {name}! You're doing great learning Python!"

In [ ]:
def encouragement(name):
  for x in name:
   print("Keep going", x + '!' + ' ' "You're doing great learning Python!")
   
 

8.2 Print out `encouragement` for each of the names in the list below

In [ ]:
student_names = ['Allison', 'Riley', 'Sarah', 'Davide', 'Lane', 'Kahri', 'Ian']
encouragement(student_names)

Keep going Allison! You're doing great learning Python!
Keep going Riley! You're doing great learning Python!
Keep going Sarah! You're doing great learning Python!
Keep going Davide! You're doing great learning Python!
Keep going Lane! You're doing great learning Python!
Keep going Kahri! You're doing great learning Python!
Keep going Ian! You're doing great learning Python!


8.3 Some of the students have expressed more interest in learning R than Python. Adjust your function to print "Keep going, {name}! You're doing great learning {language}!" based on the language the student prefers to learn, found below.

In [ ]:
prefer_r = ['Allison', 'Riley', 'Davide', 'Kahri']
prefer_python = ['Sarah', 'Lane', 'Ian']

def encouragement(name):
  for x in name:
   print("Keep going", x + '!' + ' ' "You're doing great learning R!")

encouragement(prefer_r)
   
 

Keep going Allison! You're doing great learning R!
Keep going Riley! You're doing great learning R!
Keep going Davide! You're doing great learning R!
Keep going Kahri! You're doing great learning R!


In [ ]:
prefer_python = ['Sarah', 'Lane', 'Ian']


def encouragement(name):
  for x in name:
   print("Keep going", x + '!' + ' ' "You're doing great learning Python!")
  

encouragement(prefer_python)

Keep going Sarah! You're doing great learning Python!
Keep going Lane! You're doing great learning Python!
Keep going Ian! You're doing great learning Python!


### 9. Data Structure

9.1 Create a variable called `my_list` that stores the values 1, 2, 3, 4

In [ ]:
my_list = [1, 2, 3, 4]
print(my_list)

[1, 2, 3, 4]


9.2 Add the values 6 and 7 to `my_list`

In [ ]:
my_list.append(6)
my_list

[1, 2, 3, 4, 6]

In [ ]:
my_list.append(7)
my_list

[1, 2, 3, 4, 6, 7]

9.3
Add the values 8 and 9 to my_list

In [ ]:
my_list.append(8)
my_list

[1, 2, 3, 4, 6, 7, 8]

In [ ]:
my_list.append(9)
my_list

[1, 2, 3, 4, 6, 7, 8, 9]

9.4 Remove the value [8, 9] from my_list and add the value 5 in between the value of 4 and 6

In [ ]:
my_list.remove(9)
my_list

[1, 2, 3, 4, 6, 7, 8]

In [ ]:
my_list.remove(8)
my_list

[1, 2, 3, 4, 6, 7]

9.5 Make and print a dictionary called `my_dictionry` that has the following structure:

A: Avocado 

B: Banana

C: Cheese

D: Dolla', dolla' bills y'all

In [ ]:
my_dictionary = {"A": "Avocado", "B": "Banana", "C": "Cheese", "D": "Dolla',dolla' bills y'all"}
print(my_dictionary)

{'A': 'Avocado', 'B': 'Banana', 'C': 'Cheese', 'D': "Dolla',dolla' bills y'all"}


9.6 Return the third element of `my_dictionary`

In [ ]:
print(my_dictionary.get('C'))

Cheese


9.7 What is a tuple?

A tuple is used to store multiple items in a single variable. It is one of 4 built-in data types in python used to store collections of data; tuples are ordered and unchangeable unlike other objects such as lists and dictionaries. Provided is an example of a tuple:


In [ ]:
tuple = ("Boston College", "Boston University", "UVM")
print(tuple)

('Boston College', 'Boston University', 'UVM')


9.8 What are the values of the set {1, 2, 3, 4, 4, 5, 6, 7, 7, 8}

In [ ]:
set  = {1, 2, 3, 4, 4, 5, 6, 7, 7, 8}

print(set)

      

{1, 2, 3, 4, 5, 6, 7, 8}


### 10. Loading Data

10.1 Load pandas and alias it as pd

In [ ]:
import pandas as pd

10.2 Load the `IMDB_Top_250` dataset as `imdb`

In [ ]:
from google.colab import files
import io

uploaded = files.upload()



TypeError: ignored

In [ ]:
import pandas as pd
import io

df = pd.read_csv(io.BytesIO(uploaded['IMDB_Top_250.csv']))
print(df)

     Unnamed: 0  Num                     Title  Year   Released  Runtime  \
0             0    1  The Shawshank Redemption  1994  14-Oct-94  142 min   
1             1    2             The Godfather  1972  24-Mar-72  175 min   
2             2    3    The Godfather: Part II  1974  20-Dec-74  202 min   
3             3    4           The Dark Knight  2008  18-Jul-08  152 min   
4             4    5              12 Angry Men  1957   1-Apr-57   96 min   
..          ...  ...                       ...   ...        ...      ...   
245         245  246          The Lost Weekend  1945   1-Jan-46  101 min   
246         246  247             Short Term 12  2013  23-Aug-13   96 min   
247         247  248           His Girl Friday  1940  18-Jan-40   92 min   
248         248  249        The Straight Story  1999   3-Nov-99  112 min   
249         249  250       Slumdog Millionaire  2008  25-Dec-08  120 min   

                      Genre                      Director  \
0              Crime, Dram

### 11. Summary Statistics

11.1 What are the datatypes for `imdb`?

In [ ]:
df.dtypes

Unnamed: 0      int64
Num             int64
Title          object
Year            int64
Released       object
Runtime        object
Genre          object
Director       object
Writer         object
Actors         object
Plot           object
Language       object
Country        object
Awards         object
Metascore     float64
imdbRating    float64
imdbVotes      object
imdbID         object
Type           object
DVD            object
BoxOffice      object
Production     object
Website        object
dtype: object

11.2 How many rows and columns are present in your dataset?

The IMDB_Top_250 file contains 250 rows and 23 columns.

11.3 Return the first 10 rows and last 10 rows of your dataset

In [ ]:
df.head(10)

Unnamed: 0  Num                                              Title  Year  \
0           0    1                           The Shawshank Redemption  1994   
1           1    2                                      The Godfather  1972   
2           2    3                             The Godfather: Part II  1974   
3           3    4                                    The Dark Knight  2008   
4           4    5                                       12 Angry Men  1957   
5           5    6                                   Schindler's List  1993   
6           6    7      The Lord of the Rings: The Return of the King  2003   
7           7    8                                       Pulp Fiction  1994   
8           8    9                                         Fight Club  1999   
9           9   10  The Lord of the Rings: The Fellowship of the Ring  2001   

    Released  Runtime                      Genre              Director  \
0  14-Oct-94  142 min               Crime, Drama        Frank Darabont   
1  24-Mar-72  175 min               Crime, Drama  Francis Ford Coppola   
2  20-Dec-74  202 min               Crime, Drama  Francis Ford Coppola   
3  18-Jul-08  152 min       Action, Crime, Drama     Christopher Nolan   
4   1-Apr-57   96 min               Crime, Drama          Sidney Lumet   
5   4-Feb-94  195 min  Biography, Drama, History      Steven Spielberg   
6  17-Dec-03  201 min  Adventure, Drama, Fantasy         Peter Jackson   
7  14-Oct-94  154 min               Crime, Drama     Quentin Tarantino   
8  15-Oct-99  139 min                      Drama         David Fincher   
9  19-Dec-01  178 min  Adventure, Drama, Fantasy         Peter Jackson   

                                              Writer  \
0  Stephen King (short story "Rita Hayworth and S...   
1  Mario Puzo (screenplay), Francis Ford Coppola ...   
2  Francis Ford Coppola (screenplay), Mario Puzo ...   
3  Jonathan Nolan (screenplay), Christopher Nolan...   
4  Reginald Rose (story), Reginald Rose (screenplay)   
5  Thomas Keneally (book), Steven Zaillian (scree...   
6  J.R.R. Tolkien (novel), Fran Walsh (screenplay...   
7  Quentin Tarantino (stories), Roger Avary (stor...   
8     Chuck Palahniuk (novel), Jim Uhls (screenplay)   
9  J.R.R. Tolkien (novel), Fran Walsh (screenplay...   

                                              Actors  ...  \
0  Tim Robbins, Morgan Freeman, Bob Gunton, Willi...  ...   
1  Marlon Brando, Al Pacino, James Caan, Richard ...  ...   
2  Al Pacino, Robert Duvall, Diane Keaton, Robert...  ...   
3  Christian Bale, Heath Ledger, Aaron Eckhart, M...  ...   
4  Martin Balsam, John Fiedler, Lee J. Cobb, E.G....  ...   
5  Liam Neeson, Ben Kingsley, Ralph Fiennes, Caro...  ...   
6   Noel Appleby, Ali Astin, Sean Astin, David Aston  ...   
7  Tim Roth, Amanda Plummer, Laura Lovelace, John...  ...   
8  Edward Norton, Brad Pitt, Meat Loaf, Zach Grenier  ...   
9  Alan Howard, Noel Appleby, Sean Astin, Sala Baker  ...   

                                              Awards Metascore imdbRating  \
0  Nominated for 7 Oscars. Another 19 wins & 30 n...      80.0        9.3   
1    Won 3 Oscars. Another 23 wins & 27 nominations.     100.0        9.2   
2    Won 6 Oscars. Another 10 wins & 20 nominations.      85.0        9.0   
3  Won 2 Oscars. Another 151 wins & 153 nominations.      82.0        9.0   
4  Nominated for 3 Oscars. Another 16 wins & 8 no...      96.0        8.9   
5    Won 7 Oscars. Another 71 wins & 33 nominations.      93.0        8.9   
6  Won 11 Oscars. Another 197 wins & 122 nominati...      94.0        8.9   
7     Won 1 Oscar. Another 60 wins & 68 nominations.      94.0        8.9   
8  Nominated for 1 Oscar. Another 10 wins & 31 no...      66.0        8.8   
9  Won 4 Oscars. Another 113 wins & 124 nominations.      92.0        8.8   

   imdbVotes     imdbID   Type        DVD      BoxOffice  \
0  1,825,626  tt0111161  movie  27-Jan-98            NaN   
1  1,243,444  tt0068646  movie   9-Oct-01            NaN   
2

In [ ]:
df.tail(10)

Unnamed: 0  Num                Title  Year   Released  Runtime  \
240         240  241             Deadpool  2016  12-Feb-16  108 min   
241         241  242       The Wild Bunch  1969   7-Aug-69  135 min   
242         242  243              Aladdin  1992  25-Nov-92   90 min   
243         243  244             Big Fish  2003   9-Jan-04  125 min   
244         244  245               Patton  1970   2-Apr-70  172 min   
245         245  246     The Lost Weekend  1945   1-Jan-46  101 min   
246         246  247        Short Term 12  2013  23-Aug-13   96 min   
247         247  248      His Girl Friday  1940  18-Jan-40   92 min   
248         248  249   The Straight Story  1999   3-Nov-99  112 min   
249         249  250  Slumdog Millionaire  2008  25-Dec-08  120 min   

                            Genre                      Director  \
240     Action, Adventure, Comedy                    Tim Miller   
241    Action, Adventure, Western                 Sam Peckinpah   
242  Animation, Adventure, Comedy     Ron Clements, John Musker   
243     Adventure, Drama, Fantasy                    Tim Burton   
244         Biography, Drama, War         Franklin J. Schaffner   
245              Drama, Film-Noir                  Billy Wilder   
246                         Drama         Destin Daniel Cretton   
247        Comedy, Drama, Romance                  Howard Hawks   
248              Biography, Drama                   David Lynch   
249                         Drama  Danny Boyle, Loveleen Tandan   

                                                Writer  \
240                          Rhett Reese, Paul Wernick   
241  Walon Green (screenplay), Sam Peckinpah (scree...   
242  Ron Clements (screenplay), John Musker (screen...   
243   Daniel Wallace (novel), John August (screenplay)   
244  Francis Ford Coppola (screen story and screenp...   
245  Charles R. Jackson (from the novel by), Charle...   
246                              Destin Daniel Cretton   
247  Charles Lederer (screen play), Ben Hecht (from...   
248                           John Roach, Mary Sweeney   
249   Simon Beaufoy (screenplay), Vikas Swarup (novel)   

                                                Actors  ...  \
240  Ryan Reynolds, Karan Soni, Ed Skrein, Michael ...  ...   
241  William Holden, Ernest Borgnine, Robert Ryan, ...  ...   
242  Scott Weinger, Robin Williams, Linda Larkin, J...  ...   
243  Ewan McGregor, Albert Finney, Billy Crudup, Je...  ...   
244  George C. Scott, Karl Malden, Stephen Young, M...  ...   
245  Ray Milland, Jane Wyman, Phillip Terry, Howard...  ...   
246  Brie Larson, John Gallagher Jr., Stephanie Bea...  ...   
247  Cary Grant, Rosalind Russell, Ralph Bellamy, G...  ...   
248  Sissy Spacek, Jane Galloway Heitz, Joseph A. C...  ...   
249  Dev Patel, Saurabh Shukla, Anil Kapoor, Raj Zu...  ...   

                                                Awards Metascore imdbRating  \
240  Nominated for 2 Golden Globes. Another 23 wins...      65.0        8.0   
241  Nominated for 2 Oscars. Another 5 wins & 4 nom...       NaN        8.0   
242    Won 2 Oscars. Another 25 wins & 18 nominations.       NaN        8.0   
243  Nominated for 1 Oscar. Another 1 win & 68 nomi...      58.0        8.0   
244     Won 7 Oscars. Another 17 wins & 8 nominations.      91.0        8.0   
245     Won 4 Oscars. Another 12 wins & 3 nominations.       NaN        8.0   
246                          35 wins & 70 nominations.      82.0        8.0   
247                                            2 wins.       NaN        8.0   
248  Nominated for 1 Oscar. Another 14 wins & 37 no...      86.0        8.0   
249  Won 8 Oscars. Another 144 wins & 126 nominations.      86.0        8.0   

    imdbVotes     imdbID   Type        DVD      BoxOffice  \
240   644,281  tt1431045  movie  10-May-16  $328,674,489    
241    65,987  tt0065214  movie  27-Aug-97            NaN   
242   281,651  tt0103639  movie   5-Oct-04            NaN   
243   364,938  tt0319061  movie  27-Ap

11.4 Return information about your entire dataset

In [ ]:
df.describe()

Unnamed: 0         Num         Year   Metascore  imdbRating
count  250.000000  250.000000   250.000000  177.000000  250.000000
mean   124.500000  125.500000  1982.676000   80.045198    8.244000
std     72.312977   72.312977    24.809212   11.030340    0.245735
min      0.000000    1.000000  1921.000000   53.000000    8.000000
25%     62.250000   63.250000  1961.250000   72.000000    8.100000
50%    124.500000  125.500000  1988.000000   81.000000    8.200000
75%    186.750000  187.750000  2003.000000   89.000000    8.375000
max    249.000000  250.000000  2017.000000  100.000000    9.300000

11.5 How many NaNs or missing values are in your dataset? Is there any changes that may need to be done to your dataset to handle such cases

In [ ]:
print(df.isnull().sum())

Unnamed: 0      0
Num             0
Title           0
Year            0
Released        2
Runtime         0
Genre           0
Director        0
Writer          1
Actors          0
Plot            0
Language        0
Country         0
Awards          5
Metascore      73
imdbRating      0
imdbVotes       0
imdbID          0
Type            0
DVD             3
BoxOffice     175
Production      0
Website       119
dtype: int64


11.6 Can you return basic statisitics for atleast 3-5 columns of your dataset? What can you learn from looking at these statistics?

In [ ]:
#basic stats for metascores

#imbdrating
#boxoffice


### 12. Slicing

12.1 Return the entire second column of your dataset

In [ ]:

df['Title']


3          The Dark Knight
5                      Num
6                      Num
7                      Num
8                      Num
              ...         
229      The King's Speech
234          Blood Diamond
240               Deadpool
243               Big Fish
249    Slumdog Millionaire
Name: Title, Length: 70, dtype: object

12.2 Return rows 5-10 of the second column of your dataset 

In [ ]:
import pandas as pd
import io
from google.colab import files
import io

uploaded = files.upload()

df = pd.read_csv(io.BytesIO(uploaded['IMDB_Top_250.csv']))
print(df)

df.iloc[5:10] = "Title"

Saving IMDB_Top_250.csv to IMDB_Top_250 (3).csv
     Unnamed: 0  Num                     Title  Year   Released  Runtime  \
0             0    1  The Shawshank Redemption  1994  14-Oct-94  142 min   
1             1    2             The Godfather  1972  24-Mar-72  175 min   
2             2    3    The Godfather: Part II  1974  20-Dec-74  202 min   
3             3    4           The Dark Knight  2008  18-Jul-08  152 min   
4             4    5              12 Angry Men  1957   1-Apr-57   96 min   
..          ...  ...                       ...   ...        ...      ...   
245         245  246          The Lost Weekend  1945   1-Jan-46  101 min   
246         246  247             Short Term 12  2013  23-Aug-13   96 min   
247         247  248           His Girl Friday  1940  18-Jan-40   92 min   
248         248  249        The Straight Story  1999   3-Nov-99  112 min   
249         249  250       Slumdog Millionaire  2008  25-Dec-08  120 min   

                      Genre            

In [ ]:
df.loc[5:10, 'Title']

5            Title
6            Title
7            Title
8            Title
9            Title
10    Forrest Gump
Name: Title, dtype: object

### 13. Cleaning

13.1 Remove all NaN values from the dataset. Return output that proves there are 0 NaN values

In [ ]:
df.dropna()

Unnamed: 0    Num                Title   Year   Released  Runtime  \
3            3      4      The Dark Knight   2008  18-Jul-08  152 min   
5        Title  Title                Title  Title      Title    Title   
6        Title  Title                Title  Title      Title    Title   
7        Title  Title                Title  Title      Title    Title   
8        Title  Title                Title  Title      Title    Title   
..         ...    ...                  ...    ...        ...      ...   
229        229    230    The King's Speech   2010  25-Dec-10  118 min   
234        234    235        Blood Diamond   2006   8-Dec-06  143 min   
240        240    241             Deadpool   2016  12-Feb-16  108 min   
243        243    244             Big Fish   2003   9-Jan-04  125 min   
249        249    250  Slumdog Millionaire   2008  25-Dec-08  120 min   

                          Genre                      Director  \
3          Action, Crime, Drama             Christopher Nolan   
5                         Title                         Title   
6                         Title                         Title   
7                         Title                         Title   
8                         Title                         Title   
..                          ...                           ...   
229            Biography, Drama                    Tom Hooper   
234  Adventure, Drama, Thriller                  Edward Zwick   
240   Action, Adventure, Comedy                    Tim Miller   
243   Adventure, Drama, Fantasy                    Tim Burton   
249                       Drama  Danny Boyle, Loveleen Tandan   

                                                Writer  \
3    Jonathan Nolan (screenplay), Christopher Nolan...   
5                                                Title   
6                                                Title   
7                                                Title   
8                                                Title   
..                                                 ...   
229                         David Seidler (screenplay)   
234  Charles Leavitt (screenplay), Charles Leavitt ...   
240                          Rhett Reese, Paul Wernick   
243   Daniel Wallace (novel), John August (screenplay)   
249   Simon Beaufoy (screenplay), Vikas Swarup (novel)   

                                                Actors  ...  \
3    Christian Bale, Heath Ledger, Aaron Eckhart, M...  ...   
5                                                Title  ...   
6                                                Title  ...   
7                                                Title  ...   
8                                                Title  ...   
..                                                 ...  ...   
229  Colin Firth, Helena Bonham Carter, Derek Jacob...  ...   
234  Leonardo DiCaprio, Djimon Hounsou, Jennifer Co...  ...   
240  Ryan Reynolds, Karan Soni, Ed Skrein, Michael ...  ...   
243  Ewan McGregor, Albert Finney, Billy Crudup, Je...  ...   
249  Dev Patel, Saurabh Shukla, Anil Kapoor, Raj Zu...  ...   

                                                Awards Metascore imdbRating  \
3    Won 2 Oscars. Another 151 wins & 153 nominations.      82.0        9.0   
5                                                Title     Title      Title   
6                                                Title     Title      Title   
7                                                Title     Title      Title   
8                                                Title     Title      Title   
..                                                 ...       ...        ...   
229  Won 4 Oscars. Another 103 wins & 194 nominations.      88.0        8.0   
234  Nominated for 5 Oscars. Another 8 wins & 26 no...      64.0        8.0   
240  Nominated for 2 Golden Globes. Another 23 wins...      65.0        8.0   
243  Nominated for 1 Oscar. Another 1 win & 68 nomi...      58.0        8.0   
249  Won 8 Oscars. Anot

In [ ]:
df.dropna(inplace=True)
df

Unnamed: 0    Num                Title   Year   Released  Runtime  \
3            3      4      The Dark Knight   2008  18-Jul-08  152 min   
5        Title  Title                Title  Title      Title    Title   
6        Title  Title                Title  Title      Title    Title   
7        Title  Title                Title  Title      Title    Title   
8        Title  Title                Title  Title      Title    Title   
..         ...    ...                  ...    ...        ...      ...   
229        229    230    The King's Speech   2010  25-Dec-10  118 min   
234        234    235        Blood Diamond   2006   8-Dec-06  143 min   
240        240    241             Deadpool   2016  12-Feb-16  108 min   
243        243    244             Big Fish   2003   9-Jan-04  125 min   
249        249    250  Slumdog Millionaire   2008  25-Dec-08  120 min   

                          Genre                      Director  \
3          Action, Crime, Drama             Christopher Nolan   
5                         Title                         Title   
6                         Title                         Title   
7                         Title                         Title   
8                         Title                         Title   
..                          ...                           ...   
229            Biography, Drama                    Tom Hooper   
234  Adventure, Drama, Thriller                  Edward Zwick   
240   Action, Adventure, Comedy                    Tim Miller   
243   Adventure, Drama, Fantasy                    Tim Burton   
249                       Drama  Danny Boyle, Loveleen Tandan   

                                                Writer  \
3    Jonathan Nolan (screenplay), Christopher Nolan...   
5                                                Title   
6                                                Title   
7                                                Title   
8                                                Title   
..                                                 ...   
229                         David Seidler (screenplay)   
234  Charles Leavitt (screenplay), Charles Leavitt ...   
240                          Rhett Reese, Paul Wernick   
243   Daniel Wallace (novel), John August (screenplay)   
249   Simon Beaufoy (screenplay), Vikas Swarup (novel)   

                                                Actors  ...  \
3    Christian Bale, Heath Ledger, Aaron Eckhart, M...  ...   
5                                                Title  ...   
6                                                Title  ...   
7                                                Title  ...   
8                                                Title  ...   
..                                                 ...  ...   
229  Colin Firth, Helena Bonham Carter, Derek Jacob...  ...   
234  Leonardo DiCaprio, Djimon Hounsou, Jennifer Co...  ...   
240  Ryan Reynolds, Karan Soni, Ed Skrein, Michael ...  ...   
243  Ewan McGregor, Albert Finney, Billy Crudup, Je...  ...   
249  Dev Patel, Saurabh Shukla, Anil Kapoor, Raj Zu...  ...   

                                                Awards Metascore imdbRating  \
3    Won 2 Oscars. Another 151 wins & 153 nominations.      82.0        9.0   
5                                                Title     Title      Title   
6                                                Title     Title      Title   
7                                                Title     Title      Title   
8                                                Title     Title      Title   
..                                                 ...       ...        ...   
229  Won 4 Oscars. Another 103 wins & 194 nominations.      88.0        8.0   
234  Nominated for 5 Oscars. Another 8 wins & 26 no...      64.0        8.0   
240  Nominated for 2 Golden Globes. Another 23 wins...      65.0        8.0   
243  Nominated for 1 Oscar. Another 1 win & 68 nomi...      58.0        8.0   
249  Won 8 Oscars. Anot

In [ ]:
print(df.isnull().sum())

Unnamed: 0    0
Num           0
Title         0
Year          0
Released      0
Runtime       0
Genre         0
Director      0
Writer        0
Actors        0
Plot          0
Language      0
Country       0
Awards        0
Metascore     0
imdbRating    0
imdbVotes     0
imdbID        0
Type          0
DVD           0
BoxOffice     0
Production    0
Website       0
dtype: int64


### 14. Sorting

14.1 Sort your dataset via a specific column in ascending order. What column might be best to sort your dataset by?

In [ ]:
df.sort_values("imdbVotes")

Unnamed: 0  Num                               Title  Year   Released  \
48           48   49                    Django Unchained  2012  25-Dec-12   
219         219  220                        The Avengers  2012   4-May-12   
89           89   90                       Batman Begins  2005  15-Jun-05   
25           25   26                        Interstellar  2014   7-Nov-14   
38           38   39                           Gladiator  2000   5-May-00   
..          ...  ...                                 ...   ...        ...   
34           34   35                        The Departed  2006   6-Oct-06   
23           23   24                 Saving Private Ryan  1998  24-Jul-98   
20           20   21            The Silence of the Lambs  1991  14-Feb-91   
78           78   79                Inglourious Basterds  2009  21-Aug-09   
17           17   18  Star Wars: Episode IV - A New Hope  1977  25-May-77   

     Runtime                       Genre                     Director  \
48   165 min              Drama, Western            Quentin Tarantino   
219  143 min              Action, Sci-Fi                  Joss Whedon   
89   140 min           Action, Adventure            Christopher Nolan   
25   169 min    Adventure, Drama, Sci-Fi            Christopher Nolan   
38   155 min    Action, Adventure, Drama                 Ridley Scott   
..       ...                         ...                          ...   
34   151 min      Crime, Drama, Thriller              Martin Scorsese   
23   169 min                  Drama, War             Steven Spielberg   
20   118 min      Crime, Drama, Thriller               Jonathan Demme   
78   153 min       Adventure, Drama, War  Quentin Tarantino, Eli Roth   
17   121 min  Action, Adventure, Fantasy                 George Lucas   

                                                Writer  \
48                                   Quentin Tarantino   
219  Joss Whedon (screenplay), Zak Penn (story), Jo...   
89   Bob Kane (characters), David S. Goyer (story),...   
25                   Jonathan Nolan, Christopher Nolan   
38   David Franzoni (story), David Franzoni (screen...   
..                                                 ...   
34   William Monahan (screenplay), Alan Mak, Felix ...   
23                                        Robert Rodat   
20       Thomas Harris (novel), Ted Tally (screenplay)   
78                                   Quentin Tarantino   
17                                        George Lucas   

                                                Actors  ...  \
48   Jamie Foxx, Christoph Waltz, Leonardo DiCaprio...  ...   
219  Robert Downey Jr., Chris Evans, Mark Ruffalo, ...  ...   
89   Christian Bale, Michael Caine, Liam Neeson, Ka...  ...   
25   Ellen Burstyn, Matthew McConaughey, Mackenzie ...  ...   
38   Russell Crowe, Joaquin Phoenix, Connie Nielsen...  ...   
..                                                 ...  ...   
34   Leonardo DiCaprio, Matt Damon, Jack Nicholson,...  ...   
23   Tom Hanks, Tom Sizemore, Edward Burns, Barry P...  ...   
20   Jodie Foster, Lawrence A. Bonney, Kasi Lemmons...  ...   
78   Brad Pitt, Mélanie Laurent, Christoph Waltz...  ...   
17   Mark Hamill, Harrison Ford, Carrie Fisher, Pet...  ...   

                                                Awards Metascore imdbRating  \
48    Won 2 Oscars. Another 56 wins & 151 nominations.      81.0        8.4   
219  Nominated for 1 Oscar. Another 37 wins & 79 no...      69.0        8.1   
89   Nominated for 1 Oscar. Another 14 wins & 71 no...      70.0        8.3   
25     Won 1 Oscar. Another 41 wins & 142 nominations.      74.0        8.6   
38    Won 5 Oscars. Another 53 wins & 101 nominations.      67.0        8.5   
..                                                 ...       ...        ...   
34    Won 4 Oscars. Another 92 wins & 134 nominations.      85.0        8.5   
23     Won 5 Oscars. Another 74 wins & 74 nominations.      90.0        8.6   
20     Won 5 Oscars. Another 49 wins &

14.2 If you have multiple rows with the same information or numerical data, how would you handle ties in the sorting function?

Python will sort the tied rows in the same way they were in before the sorting. A programmer can utilize the sort_index() method usng numpy.argsort() with default kind=quicksort, if you are soring only by one column. (source: https://stackoverflow.com/questions/19580900/how-is-pandas-deciding-order-in-a-sort-when-there-is-a-tie)

14.3 Try to sort your data in the following ways (there are more ways than this to understand your dataset):
        
        1. Find the top 10 least values of a specific column (make sure column has numerical data inside)
        2. Can you sort the data to find the 10 greatest values of a different column?

In [ ]:
df.sort_values('imdbID')
print(df)

     Unnamed: 0  Num                     Title  Year   Released  Runtime  \
0             0    1  The Shawshank Redemption  1994  14-Oct-94  142 min   
1             1    2             The Godfather  1972  24-Mar-72  175 min   
2             2    3    The Godfather: Part II  1974  20-Dec-74  202 min   
3             3    4           The Dark Knight  2008  18-Jul-08  152 min   
4             4    5              12 Angry Men  1957   1-Apr-57   96 min   
..          ...  ...                       ...   ...        ...      ...   
245         245  246          The Lost Weekend  1945   1-Jan-46  101 min   
246         246  247             Short Term 12  2013  23-Aug-13   96 min   
247         247  248           His Girl Friday  1940  18-Jan-40   92 min   
248         248  249        The Straight Story  1999   3-Nov-99  112 min   
249         249  250       Slumdog Millionaire  2008  25-Dec-08  120 min   

                      Genre                      Director  \
0              Crime, Dram

### 15. Custom Columns

15.1 Make a custom column in your dataset. Make sure to name it appropriately!

In [ ]:
import pandas as pd
import io
from google.colab import files
import io

uploaded = files.upload()

df = pd.read_csv(io.BytesIO(uploaded['IMDB_Top_250.csv']))
print(df)


Saving IMDB_Top_250.csv to IMDB_Top_250 (5).csv
     Unnamed: 0  Num                     Title  Year   Released  Runtime  \
0             0    1  The Shawshank Redemption  1994  14-Oct-94  142 min   
1             1    2             The Godfather  1972  24-Mar-72  175 min   
2             2    3    The Godfather: Part II  1974  20-Dec-74  202 min   
3             3    4           The Dark Knight  2008  18-Jul-08  152 min   
4             4    5              12 Angry Men  1957   1-Apr-57   96 min   
..          ...  ...                       ...   ...        ...      ...   
245         245  246          The Lost Weekend  1945   1-Jan-46  101 min   
246         246  247             Short Term 12  2013  23-Aug-13   96 min   
247         247  248           His Girl Friday  1940  18-Jan-40   92 min   
248         248  249        The Straight Story  1999   3-Nov-99  112 min   
249         249  250       Slumdog Millionaire  2008  25-Dec-08  120 min   

                      Genre            

In [ ]:
df['Soundtrack'] = pd.NaT
df

Unnamed: 0  Num                     Title  Year   Released  Runtime  \
0             0    1  The Shawshank Redemption  1994  14-Oct-94  142 min   
1             1    2             The Godfather  1972  24-Mar-72  175 min   
2             2    3    The Godfather: Part II  1974  20-Dec-74  202 min   
3             3    4           The Dark Knight  2008  18-Jul-08  152 min   
4             4    5              12 Angry Men  1957   1-Apr-57   96 min   
..          ...  ...                       ...   ...        ...      ...   
245         245  246          The Lost Weekend  1945   1-Jan-46  101 min   
246         246  247             Short Term 12  2013  23-Aug-13   96 min   
247         247  248           His Girl Friday  1940  18-Jan-40   92 min   
248         248  249        The Straight Story  1999   3-Nov-99  112 min   
249         249  250       Slumdog Millionaire  2008  25-Dec-08  120 min   

                      Genre                      Director  \
0              Crime, Drama                Frank Darabont   
1              Crime, Drama          Francis Ford Coppola   
2              Crime, Drama          Francis Ford Coppola   
3      Action, Crime, Drama             Christopher Nolan   
4              Crime, Drama                  Sidney Lumet   
..                      ...                           ...   
245        Drama, Film-Noir                  Billy Wilder   
246                   Drama         Destin Daniel Cretton   
247  Comedy, Drama, Romance                  Howard Hawks   
248        Biography, Drama                   David Lynch   
249                   Drama  Danny Boyle, Loveleen Tandan   

                                                Writer  \
0    Stephen King (short story "Rita Hayworth and S...   
1    Mario Puzo (screenplay), Francis Ford Coppola ...   
2    Francis Ford Coppola (screenplay), Mario Puzo ...   
3    Jonathan Nolan (screenplay), Christopher Nolan...   
4    Reginald Rose (story), Reginald Rose (screenplay)   
..                                                 ...   
245  Charles R. Jackson (from the novel by), Charle...   
246                              Destin Daniel Cretton   
247  Charles Lederer (screen play), Ben Hecht (from...   
248                           John Roach, Mary Sweeney   
249   Simon Beaufoy (screenplay), Vikas Swarup (novel)   

                                                Actors  ... Metascore  \
0    Tim Robbins, Morgan Freeman, Bob Gunton, Willi...  ...      80.0   
1    Marlon Brando, Al Pacino, James Caan, Richard ...  ...     100.0   
2    Al Pacino, Robert Duvall, Diane Keaton, Robert...  ...      85.0   
3    Christian Bale, Heath Ledger, Aaron Eckhart, M...  ...      82.0   
4    Martin Balsam, John Fiedler, Lee J. Cobb, E.G....  ...      96.0   
..                                                 ...  ...       ...   
245  Ray Milland, Jane Wyman, Phillip Terry, Howard...  ...       NaN   
246  Brie Larson, John Gallagher Jr., Stephanie Bea...  ...      82.0   
247  Cary Grant, Rosalind Russell, Ralph Bellamy, G...  ...       NaN   
248  Sissy Spacek, Jane Galloway Heitz, Joseph A. C...  ...      86.0   
249  Dev Patel, Saurabh Shukla, Anil Kapoor, Raj Zu...  ...      86.0   

    imdbRating  imdbVotes     imdbID   Type        DVD      BoxOffice  \
0          9.3  1,825,626  tt0111161  movie  27-Jan-98            NaN   
1          9.2  1,243,444  tt0068646  movie   9-Oct-01            NaN   
2          9.0    856,870  tt0071562  movie  24-May-05            NaN   
3          9.0  1,802,351  tt0468569  movie   9-Dec-08  $533,316,061    
4          8.9    494,215  tt0050083  movie   6-Mar-01            NaN   
..         ...        ...        ...    ...        ...            ...   
245        8.0     26,834  tt0037884  movie   6-Feb-01            NaN   
246        8.0     62,348  tt2370248  movie  14-Jan-14            NaN   
247        8.0     43,551  tt0032599  movie  28-Dec-04            NaN   
248        8.0     67,045  tt0166896  movie   7-Nov-00 

15.2 Make a column that has a category assigned to it if it is above a certain threshold. 

Rememeber from wine data, you were asked to: Make a column called quality based on if a wine has a rating above 6. Then, count the values!

### 16. Group By

16.1 Group your data by a specific column to give you insight on what your data might be showing. Are there any trends you can explain?

In [ ]:
#first grouping based on "Metascore"
score = df.groupby(['Metascore']).sum()

print(score)


                          Unnamed: 0                        Num  \
Metascore                                                         
58.0                             243                        244   
60.0                             209                        210   
62.0                             306                        308   
63.0                             344                        346   
64.0                             234                        235   
65.0                             240                        241   
66.0                              40                         41   
68.0                             226                        228   
69.0                             494                        497   
70.0                              89                         90   
71.0                             230                        232   
72.0                             124                        125   
73.0                             135                        13

16.2 Make a new dataframe that only contains the information after you have grouped your data by a certain category

In [ ]:
df2 = df.groupby(['Metascore', 'Genre']).sum()
print(df2)

                                                       Unnamed: 0  \
Metascore Genre                                                     
58.0      Adventure, Drama, Fantasy                           243   
60.0      Crime, Drama                                        209   
62.0      Action, Drama, Thriller                             114   
          Drama                                               192   
63.0      Action, Adventure, Fantasy                          204   
...                                                           ...   
94.0      Animation, Adventure, Family                         52   
95.0      Animation, Family, Fantasy                          196   
96.0      Biography, Drama, History                           157   
100.0     Adventure, Family, Fantasy                          184   
Title     Title                         TitleTitleTitleTitleTitle   

                                                              Num  \
Metascore Genre                  

### 17. Visualizations

17.1 Load matplotlib.pyplot aliased as plt

In [ ]:
import matplotlib.pyplot as plt

17.2 Histograms are one of the best ways to understand the distribution of key columns in your dataset. Make a histogram of the column you created in 15.2

17.3 A histogram is a bar chart of the count of the instances of the elements in a column of your dataset. Make a bar chart that counts the frequency of the elements in the same column you chose for 16.2

17.4 A box plot is another great way to understand details of a column in your data. Make a boxplot of the same column of your dataset. What can you learn from this chart?

17.5 Turn your bar chart from 17.3 into a pie chart

17.6 A scatter plot requires the identification of both values to place on the x and the y axis. It is a great way to visualize the relationship between two variables. Make a scatterplot using two columns in your data

17.7 Add a legend and title to your scatterplot. Change the color of the points on your scatter plot too.

17.8 Return the scatterplot from 17.7 and the pie chart from 16.5 in the same ouput by making a grid. **<font color = 'blue'> Google </font>** it if you need help!